### PR TITLE
feat: Support Python 3.10

### PR DESCRIPTION
**Pull request by henryiii**
This is testing using oldest supported numpy and enabling 3.10.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -91,7 +91,7 @@ jobs:
           else
             conda install nomkl "numpy${{ matrix.numpy-requirement }}" "scipy${{ matrix.scipy-requirement }}"
           fi
-          python -m pip install -e .[$QUTIP_TARGET]
+          python -m pip install -e .[$QUTIP_TARGET] --no-build-isolation
```
Comment (henryiii) [inline .github/workflows/tests.yml:94]: (If you don't do this, you'll get PyPI requirements when you are building your package, instead of conda ones)

**Comment by hodgestar** [comment]
@henryiii Thank you for taking the time to make this much nicer looking PR. It does look much better, I just have to try convince myself everything will work now.

It looks like the build is currently failing because `--no-build-isolation` doesn't install the build depedencies so we'll need to install them manually. I wonder what the best way to support that is (since users would likely have to do it themselves regularly too).

**Comment by henryiii** [comment]
Users should not be installing pip dependencies into a conda environment. They should be using conda, not pip. You are only running into this because you are using pip inside a conda environment, and yes, you'll need the dependencies (with their proper conda names) in that environment.

(And, since wheels work inside conda, `pip install cutip` should work inside a conda env, even if they should be using conda - it's only because you are building inside it, which is very much not supported for normal users)

File: .github/workflows/tests.yml (hunk 1 above)
Comment (henryiii) [reply .github/workflows/tests.yml:94]:
```suggestion
          conda install packaging cython
          python -m pip install -e .[$QUTIP_TARGET] --no-build-isolation
```
(And any more, if there are any others missing)

File: setup.cfg [hunk 2]
```diff
@@ -30,11 +30,6 @@ install_requires =
     numpy>=1.16.6
     scipy>=1.0
     packaging
-setup_requires =
-    numpy>=1.16.6,<1.20
-    scipy>=1.0
-    cython>=0.29.20
-    packaging
```
Comment (hodgestar) [inline setup.cfg:37]: Removing this prevents use of `python setup.py develop` and I don't think there is another way to rebuild only the Cython files that have changed while developing at the moment. :/

Comment (henryiii) [reply setup.cfg:37]: This has been deprecated for a while now, and recently has started producing warnings. I would recommend requiring users running `setup.py` to pre-install requirements. And I'd worry about using this command: https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html 

(But removing isn't important for this PR)

Comment (henryiii) [reply setup.cfg]: Put it back in, but without the problematic numpy constraints.

Comment (hodgestar) [reply setup.cfg]: Thanks!

Re the article -- yes, I'm aware of the various pip options, but `pip install -e .` doesn't cache Cython file builds (maybe there is some new intree-builds version of pip that I haven't seen yet that does?).

Comment (henryiii) [reply setup.cfg]: Have you tried this with pip 21.3? I'd have thought this might have started working now. Haven't tested it yet for this, though.

**Comment by hodgestar** [comment]
@henryiii I've been borrowing some of your changes from here and incorporating them into #1735.

File: setup.cfg (hunk 2 above)
Comment (hodgestar) [reply setup.cfg]: It looks like it does with pip 21.3 if I run with `pip install -e . --no-build-isolation -v`. Not the friendliest command, but progress?

**Comment by hodgestar** [comment]
@henryiii I'm closing this in favour of #1747, which is very similar and was inspired by this PR. Thank you for all your help! Further suggestions welcomed too.

**Pull request closed without merging**